cgsmiles/graph_layout.py

```diff
@@ -0,0 +1,222 @@
+        for edge2 in graph.edges:
+            if edge1 != edge2:
+                common_node = set(edge1).intersection(edge2)
+                if common_node:
+                    common_node = common_node.pop()  # Extract the common node
+                    remaining_nodes = list(set(edge1 + edge2) - {common_node})
+                    triplet = (remaining_nodes[0], common_node, remaining_nodes[1])
+                    triplets.add(triplet)
+    return triplets
+
+def assign_angles(graph):
+    """
+    Assign angle values for all angles in a molecule
+    graph. We assume they are all 120 if atoms have
+    2 or 3 neighbors. Otherwise, they are skipped. On
+    top we take care that fife membered rings and triangles
+    have the correct angle sum.
+
+    Parameters
+    ----------
+    graph: nx.Graph
+
+    Returns
+    -------
+    tuple([[abc.hashable,abc.hashable,abc.hashable,int])
+        tuple with node keys corresponding to the angle
+        and the reference angle value
+    """
+    triplets = find_triplets(graph)
+    angles = []
+    for triplet in triplets:
+        if graph.degree(triplet[1]) in [2, 3]:
+            angles.append((triplet[0],
+                          triplet[1],
+                          triplet[2],
+                          120))
+    return angles
+
+def assign_bonds(graph, pos):
+    """
+    Iterate over all edges and compute the distance
+    between the edges defined by the `pos` coordinate
+    array.
+
+    Parameters
+    ----------
+    graph: nx.Graph
+    pos: np.ndarray((n,2))
+
+    Returns
+    -------
+    tuple([abc.hashable, abc.hashable, float])
+        tuple of node keys and distance
+    """
+    bonds = []
+    for edge in graph.edges:
+        dist = np.linalg.norm(pos[edge[0]]-pos[edge[1]])
+        bonds.append((edge[0], edge[1], dist))
+    return bonds
+
+def angle_between(v1, v2):
+    """
+    Compute angle between two 2D vectors.
+    """
+    cos_theta = np.dot(v1, v2) / (np.linalg.norm(v1) * np.linalg.norm(v2))
+    angle_rad = np.arccos(np.clip(cos_theta, -1.0, 1.0))
+    angle_deg = np.degrees(angle_rad)
+    return angle_deg
+
+
+def bond_potential(atom_pos, ref):
+    """
+    Compute energy associated to the bonds.
+    """
+    energy = 10000*(ref-np.linalg.norm(atom_pos[1]-atom_pos[0]))**2
+    return energy
+
+def angle_potential(atom_pos, ref):
+    """
+    Compute the energy associated with an
+    angular potential.
+    """
+    v1 = atom_pos[1] - atom_pos[0]
+    v2 = atom_pos[1] - atom_pos[2]
+    angle = angle_between(v1, v2)
+    energy = (ref-angle_between(v1, v2))**2
+    return energy
+
+def nonbonded_potential(positions):
+    """
+    Compute a simple repulsive nonbonded potential
+    between all nodes whos positions are described by
+    `positions` array.
+    """
+    dist_matrix = squareform(pdist(positions, 'euclidean'))
+    dist_matrix = dist_matrix[dist_matrix > 10**-6]
+    energy = np.sum(np.power(dist_matrix, -8))
+    return energy
```

Comment:
You don't need the `squareform`, right? You only really care about the upper (or lower) triangular values.